### PR TITLE
some performance optimizations for starting apps

### PR DESCRIPTION
**Pull request by devoncarew**
Some performance optimizations for starting apps:
- only seed the random value if we're going to use it (~600ms)
- switch to the sha256 implementation from `crypto` (~700ms => 70ms)
- default to not calling stop when starting an app. Android is ok w/ it, and stop isn't implemented for iOS (stop was ~500ms)
- reduce the number of times we call `simctl list --json` - each call was ~100ms, and we had 5-6 calls. Down to 2 now. This is the `isConnected` change. The semantics of that method are different now that we only create a `Device` instance for devices that actually exist

### CODE REVIEW

**Comment by devoncarew** [comment]
The biggest outstanding piece of work now is zipping the snapshot into an flx. We can re-write to shell out to `zip` when it's available, and save about 1000ms.

**Comment by devoncarew** [comment]
flutter -v run:

```
[  17 ms] - Using Android SDK at /Users/devoncarew/tools/android-sdk-macosx/.
[  17 ms] - [/Users/devoncarew/tools/android-sdk-macosx/, SDK version 23, build-tools 23.0.2]
[  39 ms] - xcodebuild -version
[ 102 ms] - Xcode 7.2.1
Build version 7C1002
[ 105 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb devices -l
[ 115 ms] - List of devices attached
TA95000FQA             device usb:336592896X product:peregrine_retus model:XT1045 device:peregrine
[ 149 ms] - which idevice_id
[ 152 ms] - /Users/devoncarew/homebrew/bin/idevice_id
[ 152 ms] - /Users/devoncarew/homebrew/bin/idevice_id -l
[ 161 ms] - /usr/bin/xcrun simctl list --json devices
[ 242 ms] - Downloading toolchain.
[ 270 ms] - /usr/bin/defaults read /Users/devoncarew/flutter/flutter_sunflower/ios/Info CFBundleIdentifier
[ 314 ms] - io.flutter.runner.Runner
[ 315 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb devices -l
[ 325 ms] - List of devices attached
TA95000FQA             device usb:336592896X product:peregrine_retus model:XT1045 device:peregrine
[ 330 ms] - /Users/devoncarew/homebrew/bin/idevice_id -l
[ 336 ms] - /usr/bin/xcrun simctl list --json devices
[ 413 ms] - Running build command.
[ 416 ms] - APK up to date. Skipping build step.
[ 416 ms] - Running install command.
[ 417 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA shell pm path com.example.flutter_sunflower
[ 871 ms] - package:/data/app/com.example.flutter_sunflower-1/base.apk
[ 874 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA shell cat /data/local/tmp/sky.com.example.flutter_sunflower.sha1
[ 898 ms] - c5ea78f5d3d30eb1e2c19c29a8100836ce2522cc
[1156 ms] - Running build command for AndroidDevice TA95000FQA.
[1156 ms] Starting lib/main.dart on XT1045...
[1157 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb version
[1166 ms] - Android Debug Bridge version 1.0.32
Revision 09a0d98bebce-android
[1168 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA start-server
[1178 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA shell getprop ro.build.version.sdk
[1213 ms] - 22
[1291 ms] - /Users/devoncarew/projects/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/sky_snapshot /Users/devoncarew/flutter/flutter_sunflower/lib/main.dart --package-root=packages --snapshot=/var/folders/00/0w978000h01000cxqpysvccm003j4x/T/flutter_toolsZTLoaB/snapshot_blob.bin
[1684 ms] - Building /var/folders/00/0w978000h01000cxqpysvccm003j4x/T/flutter_toolsZTLoaB/app.flx
[1710 ms] - KeyPair from privatekey.der: null.
[1710 ms] - Encoding zip file.
[2543 ms] - Creating flx at /var/folders/00/0w978000h01000cxqpysvccm003j4x/T/flutter_toolsZTLoaB/app.flx.
[2592 ms] - Built and signed flx at /var/folders/00/0w978000h01000cxqpysvccm003j4x/T/flutter_toolsZTLoaB/app.flx.
[2592 ms] - Starting bundle for AndroidDevice TA95000FQA.
[2594 ms] - AndroidDevice TA95000FQA startBundle
[2594 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA forward tcp:8181 tcp:8181
[2606 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA -s TA95000FQA logcat -c
[2645 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA push /var/folders/00/0w978000h01000cxqpysvccm003j4x/T/flutter_toolsZTLoaB/app.flx /data/local/tmp/dev.flx
[2787 ms] - /Users/devoncarew/tools/android-sdk-macosx/platform-tools/adb -s TA95000FQA shell am start -a android.intent.action.RUN -d /data/local/tmp/dev.flx -f 0x20000000 --ez enable-background-compilation true --ez enable-checked-mode true com.example.flutter_sunflower/org.domokit.sky.shell.SkyActivity
[3411 ms] - Starting: Intent { act=android.intent.action.RUN dat=/data/local/tmp/dev.flx flg=0x20000000 cmp=com.example.flutter_sunflower/org.domokit.sky.shell.SkyActivity (has extras) }
Warning: Activity not started, its current task has been brought to the front
[3412 ms] - Finished run command.
```

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 1]
```diff
@@ -238,12 +240,7 @@ class IOSSimulator extends Device {
     }
   }
 
-  @override
-  bool isConnected() {
-    if (!Platform.isMacOS)
-      return false;
-    return SimControl.instance.getConnectedDevices().any((SimDevice device) => device.udid == id);
-  }
+  bool isConnected() => Platform.isMacOS;
```
Comment (abarth) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: Should this be a getter?

**Comment by abarth** [comment]
LGTM

File: packages/flutter_tools/lib/src/ios/simulators.dart (hunk 1 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: I can switch it. It's inherited from the parent class.

Comment (devoncarew) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: Actually, I think I'll remove the method in a follow-on PR. It doesn't really make sense any more. We used to create devices w/o knowing whether they existed - we don't do that anymore.

**Pull request merged**